### PR TITLE
intial

### PR DESCRIPTION
**Pull request by aaishwarymishra**

### CODE REVIEW

File: examples/string-reverse-fixed/trainer.py [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+        self.inference_tokenizer = getattr(train_ds, "tokenizer")
+
+        # Validate active loaders
+        self._validate_dataloader_for_inference(self.train_loader, "Train")
+        if self.val_loader is not None:
+            self._validate_dataloader_for_inference(self.val_loader, "Val")
+        if self.test_loader is not None:
+            self._validate_dataloader_for_inference(self.test_loader, "Test")
+
+    def _log_inference(
```
Comment (aaishwarymishra) [inline examples/string-reverse-fixed/trainer.py:562]: Here is updated inference method

File: examples/string-reverse-fixed/trainer.py [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+            f"Epoch {engine.state.epoch}: Running inference on {name} samples..."
+        )
+
+        eos_token_id = getattr(tokenizer, "eos_token_id", None)
+
+        dataset = loader.dataset
+        if not isinstance(dataset, Sized):
+            raise TypeError(f"{name} dataset must be Sized to run inference logging.")
+        total_samples = len(dataset)
+        num_samples = min(self.inference_num_samples, total_samples)
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/trainer.py:588]: If you set `self.inference_num_samples` to a small number like 5 then it won't be able to evaluate the accuracy on the entire test set but rather only on the basis of those 5 examples? And, if you set it to the exact number of  the test dataset then you will get printed results on the entire dataset rather than a few examples?

**Comment by TahaZahid05** [comment]
@aaishwarymishra My problem with current approach is that we are trying to use the same function to perform inference on a few samples and then also reporting the metrics on the entire set which was not the purpose of this function really. I am suggesting we keep the `_log_inference` as is and write a custom function in script to report the metrics instead? This way we can print samples using `_log_inference` and have custom function to report metrics? This is code i generated for reference:

```python
def run_autoregressive_test_eval(engine: Engine) -> None:
    epoch = engine.state.epoch
    
    # Run every 5 epochs, at val peaks, or at the last training epoch
    val_metrics = trainer.val_evaluator.state.metrics
    current_val_exact = val_metrics.get("exact_accuracy", -1.0)
    prev_best_val = store.get("anchor_val_exact_acc", -1.0)
    is_new_peak = current_val_exact > prev_best_val
    is_last_epoch = (epoch == trainer.epochs)
    
    if epoch % 5 != 0 and not is_new_peak and not is_last_epoch:
        return

    trainer.model.eval()

    dataset = trainer.test_loader.dataset
    if hasattr(dataset, "dataset"):
        underlying_ds = dataset.dataset
        tokenizer = underlying_ds.tokenizer
    else:
        underlying_ds = dataset
        tokenizer = dataset.tokenizer

    # 1. Retrieve all prompts/targets from the test dataset
    prompts = [dataset[i]["source_text"] for i in range(len(dataset))]
    targets = [dataset[i]["target_text"] for i in range(len(dataset))]

    # 2. Run batched generation on the full test set (no print statements)
    predictions = []
    batch_size = 128
    
    import torch
    with torch.no_grad():
        for i in range(0, len(prompts), batch_size):
            batch_prompts = prompts[i : i + batch_size]
            L = len(batch_prompts[0])
            batch_preds = trainer.model.generate(
                batch_prompts,
                max_new_tokens=L + 1,
                tokenizer=tokenizer,
            )
            predictions.extend(batch_preds)

    # 3. Compute the true exact accuracy over the entire dataset
    total_correct = sum(1 for p, t in zip(predictions, targets) if p == t)
    test_exact_acc = total_correct / len(targets)

    # 4. Overwrite the metric in the engine state with the true test accuracy
    engine.state.metrics["inference_exact_accuracy"] = test_exact_acc
    print(f"Epoch {epoch} - Test Set Autoregressive Exact Accuracy (Dataset-wide): {test_exact_acc:.4f}")


# Registration inside run_single_experiment():
original_attach = trainer._attach_handlers

def _patched_attach() -> None:
    original_attach()
    if trainer.test_loader is not None:
        trainer.engine.add_event_handler(
            Events.EPOCH_COMPLETED,
            run_autoregressive_test_eval,
        )
    trainer.engine.add_event_handler(
        Events.EPOCH_COMPLETED,
        _make_anchor_snapshot_callback(trainer, store),
    )

trainer._attach_handlers = _patched_attach
```

File: examples/string-reverse-fixed/trainer.py (hunk 2 above)
Comment (aaishwarymishra) [reply examples/string-reverse-fixed/trainer.py:588]: auto regressive is super expensive task it should be done on limited number of examples.

**Comment by aaishwarymishra** [comment]
> @aaishwarymishra My problem with current approach is that we are trying to use the same function to perform inference on a few samples and then also reporting the metrics on the entire set which was not the purpose of this function really. I am suggesting we keep the `_log_inference` as is and write a custom function in script to report the metrics instead? This way we can print samples using `_log_inference` and have custom function to report metrics? This is code i generated for reference:
> 
> ```python
> def run_autoregressive_test_eval(engine: Engine) -> None:
>     epoch = engine.state.epoch
>     
>     # Run every 5 epochs, at val peaks, or at the last training epoch
>     val_metrics = trainer.val_evaluator.state.metrics
>     current_val_exact = val_metrics.get("exact_accuracy", -1.0)
>     prev_best_val = store.get("anchor_val_exact_acc", -1.0)
>     is_new_peak = current_val_exact > prev_best_val
>     is_last_epoch = (epoch == trainer.epochs)
>     
>     if epoch % 5 != 0 and not is_new_peak and not is_last_epoch:
>         return
> 
>     trainer.model.eval()
> 
>     dataset = trainer.test_loader.dataset
>     if hasattr(dataset, "dataset"):
>         underlying_ds = dataset.dataset
>         tokenizer = underlying_ds.tokenizer
>     else:
>         underlying_ds = dataset
>         tokenizer = dataset.tokenizer
> 
>     # 1. Retrieve all prompts/targets from the test dataset
>     prompts = [dataset[i]["source_text"] for i in range(len(dataset))]
>     targets = [dataset[i]["target_text"] for i in range(len(dataset))]
> 
>     # 2. Run batched generation on the full test set (no print statements)
>     predictions = []
>     batch_size = 128
>     
>     import torch
>     with torch.no_grad():
>         for i in range(0, len(prompts), batch_size):
>             batch_prompts = prompts[i : i + batch_size]
>             L = len(batch_prompts[0])
>             batch_preds = trainer.model.generate(
>                 batch_prompts,
>                 max_new_tokens=L + 1,
>                 tokenizer=tokenizer,
>             )
>             predictions.extend(batch_preds)
> 
>     # 3. Compute the true exact accuracy over the entire dataset
>     total_correct = sum(1 for p, t in zip(predictions, targets) if p == t)
>     test_exact_acc = total_correct / len(targets)
> 
>     # 4. Overwrite the metric in the engine state with the true test accuracy
>     engine.state.metrics["inference_exact_accuracy"] = test_exact_acc
>     print(f"Epoch {epoch} - Test Set Autoregressive Exact Accuracy (Dataset-wide): {test_exact_acc:.4f}")
> 
> 
> # Registration inside run_single_experiment():
> original_attach = trainer._attach_handlers
> 
> def _patched_attach() -> None:
>     original_attach()
>     if trainer.test_loader is not None:
>         trainer.engine.add_event_handler(
>             Events.EPOCH_COMPLETED,
>             run_autoregressive_test_eval,
>         )
>     trainer.engine.add_event_handler(
>         Events.EPOCH_COMPLETED,
>         _make_anchor_snapshot_callback(trainer, store),
>     )
> 
> trainer._attach_handlers = _patched_attach
> ```

We should not do evaluation on dataset. its supper expensive.

File: examples/string-reverse-fixed/models/transformer.py [hunk 3]
```diff
@@ -265,13 +265,13 @@ def generate(
...
         prompt_len = input_ids.size(1)
-        print(f"max_new_tokens: {max_new_tokens}, prompt_len: {prompt_len}")
+
         for _ in range(max_new_tokens):
             logits = self(generated)
             next_token_logits = logits[:, -1, :]
             next_token = torch.argmax(next_token_logits, dim=-1, keepdim=True)
             if eos_token_id is not None:
-                eos_mask = generated[:, -1:].eq(eos_token_id)
+                eos_mask = generated[:, -1].eq(eos_token_id)
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/models/transformer.py]: wasn't this a bug you found? why revert?

Comment (aaishwarymishra) [reply examples/string-reverse-fixed/models/transformer.py]: my bad I kinda reverted changes so it came back.

Other things look fine to you?

**Comment by aaishwarymishra** [comment]
Also got new lpatop so pre-commit not set, I will update soon

File: examples/string-reverse-fixed/experiment_script.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+        store["train_token_acc"] = train_metrics["token_accuracy"]
+        store["val_loss"] = val_metrics["loss"]
+        store["val_exact_acc"] = val_metrics["exact_accuracy"]
+        store["val_token_acc"] = val_metrics["token_accuracy"]
+        store["inf_exact_acc"] = engine.state.metrics.get(
+            "inference_exact_accuracy", float("nan")
+        )
+        store["inf_token_acc"] = engine.state.metrics.get(
+            "inference_token_accuracy", float("nan")
+        )
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]: can be removed as no longer being logged

File: examples/string-reverse-fixed/experiment_script.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+    x_vals: list[int] = []
+    _METRICS = [
+        "train_loss",
+        "train_exact_acc",
+        "train_token_acc",
+        "val_loss",
+        "val_exact_acc",
+        "val_token_acc",
+        "inf_exact_acc",
+        "inf_token_acc",
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]: here as well.

File: examples/string-reverse-fixed/experiment_script.py [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+            cfg = ProjectConfig.model_validate(base_cfg.model_dump(by_alias=True))
+            cfg.seed = base_cfg.seed + run_idx
+            print(f"  Run {run_idx + 1}/{NUM_RUNS}  (seed = {cfg.seed})", end="")
+
+            snap, elapsed, epochs = run_single_experiment(cfg, time_limit_seconds)
+
+            print(
+                f"  -> {elapsed:.0f}s  {epochs} epochs  "
+                f"v_acc={snap.get('val_exact_acc', float('nan')):.3f}  "
+                f"inf_acc={snap.get('inf_exact_acc', float('nan')):.3f}  "
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]: same

File: examples/string-reverse-fixed/experiment_script.py [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+    out_path: Path,
+) -> None:
+    """6-panel figure (3×2) with error bars (± 1 std)."""
+    specs: list[tuple[str, str, str | None, list[float] | None]] = [
+        ("Token Accuracy (anchored)", "train_token_acc", "val_token_acc", [0, 1]),
+        ("Exact Match Accuracy (anchored)", "train_exact_acc", "val_exact_acc", [0, 1]),
+        ("Loss (anchored)", "train_loss", "val_loss", None),
+        ("Epochs Trained", "epochs_trained", None, None),
+        ("Inference Token Acc (anchored)", "inf_token_acc", None, [0, 1]),
+        ("Inference Exact Match Acc (anchored)", "inf_exact_acc", None, [0, 1]),
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]: change to test_ar names.

File: examples/string-reverse-fixed/experiment_script.py [hunk 8]
```diff
@@ -0,0 +1,466 @@
...
+                    x_values,
+                    y_mean,
+                    yerr=y_std,
+                    color="#2ca02c",
+                    width=bar_width,
+                    capsize=3,
+                    alpha=0.8,
+                    label="Epochs",
+                )
+            elif key.startswith("inf_"):
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]: change.

File: examples/string-reverse-fixed/experiment_script.py [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+
+        store["anchor_val_exact_acc"] = current_val_exact
+        store["anchor_epoch"] = float(engine.state.epoch)
+        store["train_loss"] = train_metrics["loss"]
+        store["train_exact_acc"] = train_metrics["exact_accuracy"]
+        store["train_token_acc"] = train_metrics["token_accuracy"]
+        store["val_loss"] = val_metrics["loss"]
+        store["val_exact_acc"] = val_metrics["exact_accuracy"]
+        store["val_token_acc"] = val_metrics["token_accuracy"]
+        # No old inference metrics stored here since we run autoregressive evaluation at the end of testing.
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py]:
```suggestion
```

File: examples/string-reverse-fixed/experiment_script.py [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+        ("Exact Match Accuracy (anchored)", "train_exact_acc", "val_exact_acc", [0, 1]),
+        ("Loss (anchored)", "train_loss", "val_loss", None),
+        ("Epochs Trained", "epochs_trained", None, None),
+        ("Test Autoregressive Token Acc", "test_ar_token_acc", None, [0, 1]),
+        (
+            "Test Autoregressive Exact Match Acc",
+            "test_ar_exact_match_acc",
+            None,
+            [0, 1],
+        ),
```
Comment (TahaZahid05) [inline examples/string-reverse-fixed/experiment_script.py:257]:
```suggestion
        ("Test Autoregressive Exact Match Acc", "test_ar_exact_match_acc", None, [0, 1]),
```

**Comment by TahaZahid05** [comment]
Closing since not relevant anymore.
